### PR TITLE
Added skills description page

### LINKED ISSUE #1925: Add a page listing all die skills
There should be a page that lists all the supported die skills, their symbols and their descriptions.

(Created in response to #1924.)

### PR DESCRIPTION
**Pull request by blackshadowshade**
Fixes #1925.

This builds on the code of @AdmiralJota and the tweaks by @dwvanstone.

### CODE REVIEW

**Comment by blackshadowshade** [comment]
Here's a (zoomed-out) screenshot of what I'm seeing:

skills_page

**Comment by dwvanstone** [comment]
OOOOH! I'm very excited about this page. Thank you for fulfilling all of our dreams!

**Comment by TheOrgg** [comment]
I assume this'll be updated as new abilities are functional?

**Comment by blackshadowshade** [comment]
Die skills will be auto-updated, since the code reads from the list of skills in BMSkill.

Die types (like Wildcard) need to be added explicitly to the code.

However, all changes in skill/type descriptions will update automatically, since this code loads the descriptions from the base BMSkill* and BMDie* classes.

**Comment by cgolubi1** [comment]
Okay, so there's definitely a test thing here: if you add new APIs, you have to add responder tests for the API, and the JS tests should use the dummy results that come from the responder tests, not hardcode their own responses.  The purpose of that is to thread together those things in an integration-test-y way, so that if we change the API we'll know, and so that if we change the API in a way that breaks the UI, we'll know that too.

If you've never done this before, say something, and i'll try to come up with a better template for it, or at least a recent-ish PR that does it.

**Comment by cgolubi1** [comment]
Anyway, let's also have a dev site --- check it out folks: http://temporary.dev.buttonweavers.com/ui/skills.html

**Comment by blackshadowshade** [comment]
If I understand correctly, you're referring to the hardcoded set of Api.dieSkills and Api.dieTypes. I can see now how that circumvents responder testing.

If you wouldn't mind referring me to a relevant PR, I'll see what I can do to put that right.

**Comment by blackshadowshade** [comment]
@AdmiralJota, take a look at this and tell us what you think!

**Comment by cgolubi1** [comment]
Sure, the last API i added was for password reset, in #2509 - the relevant set of files changed in that PR is:

```
src/api/DummyApiResponder.php
test/src/api/responder00Test.php
test/src/ui/index.html
test/src/ui/js/test_ForgotPassword.js
test/src/ui/js/test_VerifyReset.js
test/src/ui/phantom-index.html
```

(Some of those you already have covered, but definitely not the first two, and the JS tests should show using the dummy data from the responder tests.)

**Comment by dwvanstone** [comment]
The dev site looks fantastic. Well done!

**Comment by TheOrgg** [comment]
Looks good to me, too!

**Comment by blackshadowshade** [comment]
@cgolubi1, I've added API tests for die skills and types.

The last thing remaining that I'm stuck on is how to acquire the API response in test_Skills.js to fill Api.dieSkills and Api.dieTypes. How do I do this?

**Comment by irilyth** [comment]
Did we want to separate skills and types in some visual way?

**Comment by blackshadowshade** [comment]
So, we've separated them out visually by using the brackets.

If we want separate tables, that would be an extra visual distinction, but I'm not sure it's strictly necessary---or helpful.

**Comment by blackshadowshade** [comment]
It *might* be possible to add a line break or something equivalent between the die skills and types. @AdmiralJota, how would we go about doing that in a clean way?

File: src/api/DummyApiResponder.php [hunk 1]
```diff
@@ -413,6 +413,19 @@ protected function get_interface_response_editForumPost($args) {
     // End of Forum-related methods
     ////////////////////////////////////////////////////////////
 
+    protected function get_interface_response_loadDieSkillsData($args) {
+        return $this->load_json_data_from_file(
+            'loadDieSkillsData',
+            $args['threadId'] . '.json'
+        );
+    }
+
```
Comment (cgolubi1) [inline src/api/DummyApiResponder.php:422]: Too much copy-paste.  There isn't going to be a 'threadId' arg, so this is going to give you a file with a useless name.  Try copying from an API that doesn't take args, e.g. `get_interface_response_loadOpenGames` is a good one to parrot.

**Comment by cgolubi1** [comment]
shadowshade: with the caveat that i haven't done this in awhile, i *think* the `getOpenGames` API method is going to continue to be a good thing to parrot.  For instance, `test_OpenGames.getOpenGames` *looks* to me like it calls the API method, and then does stuff with the results.

The way this stuff works is the tests are calling a real API, just using the endpoint dummy_responder rather than responder, and which is handing you results from JSON files, or an error if there's no such JSON file (which is why you need to use a sensible JSON filename in DummyApiResponder.php), rather than invoking PHP logic.

**Comment by blackshadowshade** [comment]
Okay, I've connected up the API. Now, I'm having trouble testing because I don't know how to get my tests to run after the page load and associated async calls have completed.

Ideally, I'd like something like:

    stop();
    Skills.showLoggedInPage();
    waitUntilPageLoadIsFinished();
    runMyTests();
    start();

I've tried

    Env.callAsyncInParallel([
      Skills.showLoggedInPage,
    ], function() {
      runMyTests();
    });

but this hangs after running Skills.showLoggedInPage() and never proceeds to the callback.

How do I get this working?

**Comment by cgolubi1** [comment]
Compare:

```js
# OpenGames.js
OpenGames.showLoggedInPage = function() {
  // Get all needed information, then display Open Games page
  OpenGames.getOpenGames(OpenGames.showPage);
};

OpenGames.getOpenGames = function(callback) {
  Env.callAsyncInParallel([
    Api.getOpenGamesData,
    { 'func': Api.getButtonData, 'args': [ null ] },
  ], callback);
};

# Skills.js
Skills.showLoggedInPage = function() {
  Env.callAsyncInParallel([
    Api.getDieSkillsData,
    Api.getDieTypesData,
  ], Skills.showPage);
};
```

If you implement Skills.js the way OpenGames.js is, with an intermediate getter function that takes showPage as a callback, then you'll be able to do the testing the same way.  LMK if i'm missing something; i might be, it's been awhile.

**Comment by blackshadowshade** [comment]
I've got the following now:

    Skills.showLoggedInPage = function() {
      // Get all needed information, then display Skills page
      Skills.getInfo(Skills.showPage);
    };

    Skills.getInfo = function(callback) {
      Env.callAsyncInParallel([
        Api.getDieSkillsData,
        Api.getDieTypesData,
      ], callback);
    };

That doesn't solve the problem, though.

For example, here's the first test:

  test("test_Skills.showLoggedInPage", function(assert) {
    stop();
    BMTestUtils.setupFakeLogin();

    Skills.showLoggedInPage();
  //  setTimeout(function() {

      var item = document.getElementById('skills_page');
      assert.equal(item.nodeName, "DIV",
        "#skills_page is a div after showLoggedInPage() is called");

      BMTestUtils.cleanupFakeLogin();
      start();
  //  }, 1000);
});

It passes with the timeout, but fails without it, with this error message: "All test functions MUST complete jQuery activity before exiting", expected 0, result 2. That's obviously due to the two async requests sent for die skill and type info.

There's still something missing, but I don't know what.

**Comment by cgolubi1** [comment]
Hmm, does the comment on the analogous `test_OpenGames.showLoggedInPage` help?

**Comment by blackshadowshade** [comment]
Hmmm, so we're not actually testing the full functionality of showLoggedInPage, just the flow of what goes on inside it? I can do that.

**Comment by cgolubi1** [comment]
> Hmmm, so we're not actually testing the full functionality of showLoggedInPage, just the flow of what goes on inside it?

Yep, that's right.

**Comment by blackshadowshade** [comment]
In that case, the tests should now be complete.

**Comment by AdmiralJota** [comment]
> It _might_ be possible to add a line break or something equivalent between the die skills and types. @AdmiralJota, how would we go about doing that in a clean way?

I'm fine with the way it is, but if you wanted to split them, the easiest way would probably be to change the Skills.directoryDiv() function so it took two arguments: headingText and displayList. Then change line 64 to `directoryDiv.append($('<h2>', { headingText }));)` and remove lines 66-70.

Then when we call Skills.directoryDiv on line 51, call it twice: once with `Skills.directoryDiv('Die Skills', Api.dieSkills.list)` and once with `Skills.directoryDiv('Die Types', Api.dieTypes.list)`.

**Comment by blackshadowshade** [comment]
Makes sense, @AdmiralJota. So, should I make this change to split the codes up into two tables? Or does that just make the display more complicated?

**Comment by irilyth** [comment]
I think it's ok as it is, but splitting it up would be an improvement.

**Comment by blackshadowshade** [comment]
Right, we're going to leave it as-is for this pull request. If anyone expresses strong feelings about this after release, we can always adjust the look and feel then.

File: src/api/ApiResponder.php [hunk 2]
```diff
@@ -22,10 +22,13 @@ class ApiResponder {
...
-     * For now, all game functionality should require login: only
-     * add things to this list if they are necessary for user
-     * creation and/or login.
+     * Functions which allow access by unauthenticated users.
+     *
+     * All game functionality should require login.
+     *
+     * Only add things to this list if they are necessary for user
+     * creation and/or login, or for any content not requiring
+     * login, like help.
```
Comment (cgolubi1) [inline src/api/ApiResponder.php:31]: You say somewhere deeper in the code that help functionality should be designed not to make database calls.  Can you say that in this comment too, as a reminder to Future Us to limit DB-using functionality accessible to non-logged-in users?

**Comment by cgolubi1** [comment]
My read of the comments is that everyone likes or is willing to tolerate this the way it is.

I want to flag that you're adding a new set of APIs which is accessible without login; i'm fine with it, but as i said in my single comment, i think we should be more explicit in ApiResponder about what the guiding principles are, for the future.  If they're "only API calls which A. don't access the DB, or B. which are necessary to get to login, are visible without logging in," i am fine with that principle, and we should say it explicitly in one place in a comment (and i think ApiResponder is the right place).

Want to modify that comment and then rebase this down to one commit?

**Comment by blackshadowshade** [comment]
Comment updated, rebase done.

**Comment by cgolubi1** [comment]
Thanks, merging.  We definitely have enough new stuff to do a release, so the next time i work on BM, i'll get started on that.

**Pull request merged**